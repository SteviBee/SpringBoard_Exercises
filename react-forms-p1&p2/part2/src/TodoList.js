import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";
import { v4 as uuid } from "uuid";

const TodoList = () => {
    // Set state initally to an empty array, state is an array of objs in the list
    const [items, setItems] = useState([])

    // Create Remove function:
    const removeTodo = evt => {
        // Shift Opt F = to auto space
        let answer = items.filter(item => item.id !== evt.target.parentElement.id)
        // Answer key pretty slick: double arrow function
        // setTodos(todos => todos.filter(todo => todo.id !== id));
        setItems(answer)
    }

    // Create Rendering function: return new array with all the item parameters
    const renderItems = () => {
        return (        
            <div className="TodoList-Todos">
                {items.map(item => (
                    <Todo 
                    key={item.id} 
                    id={item.id} 
                    todo={item.todo} 
                    removeTodo={removeTodo} 
                    update={update}>
                    </Todo>
                ))}
            </div>      
        )
    }
    // update the state obj with a todo that is updated with the new text IF the Ids match
    const update = (id, updatedTask) => {
        setItems(todos =>            
        todos.map(todo => 
            todo.id === id ? { ...todo, todo: updatedTask } : todo
            )
        );
    };

    /** Add new item object to boxlist. */
    const addItem = item => {
        // Creat newItem object with old obejects PLUS ID PLUS the key:value pairs from prior
        let newItem = { ...item, id: uuid() };
        setItems(items => [...items, newItem]);
    };

    // Return everything
    return (
        <div className="TodoList">
            <NewTodoForm addItem={addItem}/>
            {renderItems()}

        </div>
    )

}


export default TodoList;