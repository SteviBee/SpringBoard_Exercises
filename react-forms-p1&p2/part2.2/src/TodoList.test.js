import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";
// NOTE - 5/18 - had to install jest-dom via npm install jest-dom
// Also imported it here
import '@testing-library/jest-dom'

it("renders withoutcrashing", function () {
    render(<TodoList />)
})

it('matches snapshot', () => {
    const { asFragment } = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot()
});

// Critical business logic tested here: that the form displays, renders, and removes
it('should add a new box', () => {
    const { queryByText, getByLabelText } = render(<TodoList/>)
    const input = getByLabelText("Todo:")
    const btn = queryByText("Add Todo");

    expect(queryByText('Todo: Not Here')).not.toBeInTheDocument();

    fireEvent.change(input, { target: { value: 'testingIsGreat' } });
    fireEvent.click(btn);
    expect(queryByText('testingIsGreat')).toBeInTheDocument();
});

it('should REMOVE a new box', () => {
    const { queryByText, getByLabelText } = render(<TodoList/>)
    const input = getByLabelText("Todo:")
    const btn = queryByText("Add Todo");


    expect(queryByText('Todo: Not Here')).not.toBeInTheDocument();

    fireEvent.change(input, { target: { value: 'testingIsGreat2' } });
    fireEvent.click(btn);
    expect(queryByText('testingIsGreat2')).toBeInTheDocument();

    // Now remove that box
    const remove = queryByText("X")
    fireEvent.click(remove);
    expect(queryByText('testingIsGreat2')).not.toBeInTheDocument();
});

