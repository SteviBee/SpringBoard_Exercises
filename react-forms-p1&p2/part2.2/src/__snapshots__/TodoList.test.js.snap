// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form>
      <label
        for="todo"
      >
        Todo:
      </label>
      <input
        id="todo"
        name="todo"
        type="text"
        value=""
      />
      <button>
        Add Todo
      </button>
    </form>
    <div
      class="TodoList-Todos"
    />
  </div>
</DocumentFragment>
`;
